### PR TITLE
Kma 160866959 change sm hhg homepage by status

### PR DESCRIPTION
**Pull request by kimallen**
## Description

*This changes some of the text plus the status timeline state and dates based on the most current move dates (actual || pm_survey || original/requested). 

*Modifies the calculateMoveDatesFromShipment function so that it utilizes OriginalPackDate and OriginalDeliveryDate

## Reviewer Notes
Please take a look at the logic for status changes- it should cover everything, but I may have missed something.
@chrisgilmerproj  is working on a PR to make sure that all Shipments have `OriginalPackDate` and `OriginalDeliveryDate`.  
Deeper client side unit tests and e2e tests for status changes on timeline are in another story.
Expected behavior for edge cases for weird date scenarios are found in the Pivotal story.
There are several changes in text based on status- see story.

## Setup
Run `make db_dev_reset && make db_dev_migrate && make db_populate_e2e`. Log in as tsp in tsplocal.  Modify the planned dates on the shipment flow.  (The test data is not realistic), so dates need to be modified). Go to that SM on localhost.  Adjust dates again on TSP to see changes in view.  Also you can change dates in db to see changes and enter actual dates.

## Code Review Verification Steps

* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @willowbl00
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/160866959) for this change
[Pivotal story for some of the tests](https://www.pivotaltracker.com/story/show/161353928)

## Screenshots

screen shot 2018-10-30 at 5 49 38 pm

Edge case if most current delivery date is before actual pickup and pickup is before pack date:
screen shot 2018-10-30 at 2 36 42 pm

### CODE REVIEW

File: pkg/handlers/internalapi/move_dates.go [hunk 1]
```diff
@@ -76,29 +77,64 @@ func calculateMoveDates(db *pop.Connection, planner route.Planner, moveID uuid.U
...
+	pickupDates := createFutureMoveDates(mostCurrentPickupDate, 1, false, usCalendar)
 
 	firstPossibleTransitDay := time.Time(pickupDates[len(pickupDates)-1]).AddDate(0, 0, 1)
-	if shipment.EstimatedTransitDays == nil {
-		return MoveDatesSummary{}, errors.New("Shipment must have EstimatedTransitDays")
-	}
-	transitDates := createFutureMoveDates(firstPossibleTransitDay, int(*shipment.EstimatedTransitDays), true, usCalendar)
+	//if shipment.EstimatedTransitDays == nil {
+	//	return MoveDatesSummary{}, errors.New("Shipment must have EstimatedTransitDays")
+	//}
```
Comment (chrisgilmerproj) [inline pkg/handlers/internalapi/move_dates.go]: Should this be removed?

File: pkg/handlers/internalapi/move_dates.go [hunk 2]
```diff
@@ -76,29 +77,64 @@ func calculateMoveDates(db *pop.Connection, planner route.Planner, moveID uuid.U
...
 }
 
+// calculateMoveDatesFromShipment takes stored values on the shipment to calculate the most up-to-date move date ranges
+// this is used to display date ranges for the SM HHG review page and the status timeline on the post-hhg-submission landing page
 func calculateMoveDatesFromShipment(shipment *models.Shipment) (MoveDatesSummary, error) {
 	usCalendar := handlers.NewUSCalendar()
 
+	if shipment.OriginalPackDate == nil {
+		return MoveDatesSummary{}, errors.New("Shipment must have a OriginalPackDate")
+	}
```
Comment (chrisgilmerproj) [inline pkg/handlers/internalapi/move_dates.go]: What happens when you get to the review page and `OriginalPackDate` is missing? Does it just show a `*` or `missing` on the page?

File: pkg/handlers/internalapi/move_dates.go [hunk 3]
```diff
@@ -76,29 +77,64 @@ func calculateMoveDates(db *pop.Connection, planner route.Planner, moveID uuid.U
...
+	if shipment.ActualPickupDate != nil {
+		mostCurrentPickupDate = *shipment.ActualPickupDate
+	} else if shipment.PmSurveyPlannedPickupDate != nil {
+		mostCurrentPickupDate = *shipment.PmSurveyPlannedPickupDate
+	} else {
+		mostCurrentPickupDate = *shipment.RequestedPickupDate
+	}
+	if shipment.OriginalDeliveryDate == nil {
+		return MoveDatesSummary{}, errors.New("Shipment must have a OriginalDeliveryDate")
+	}
```
Comment (chrisgilmerproj) [inline pkg/handlers/internalapi/move_dates.go]: Same question as above, what happens if this is missing?

Also, for code readability it would be nice to put all of these error checks up at the top of the function.

File: pkg/handlers/internalapi/move_dates.go [hunk 4]
```diff
@@ -137,3 +173,29 @@ func createPastMoveDates(startDate time.Time, numDays int, includeWeekendsAndHol
 
 	return dates
 }
+
+func createValidDatesBetweenTwoDates(startDate time.Time, endDate time.Time, includeWeekendsAndHolidays bool, allowEarlierOrSameEndDate bool, calendar *cal.Calendar) ([]time.Time, error) {
```
Comment (chrisgilmerproj) [inline pkg/handlers/internalapi/move_dates.go]: Is it possible to use this function to remove `createFutureMoveDates` or `createPastMoveDates` or do they have different functions? And if they can't be replaced could you add function comments so it's more clear what they are doing?

File: pkg/handlers/internalapi/move_dates.go [hunk 5]
```diff
@@ -76,29 +77,64 @@ func calculateMoveDates(db *pop.Connection, planner route.Planner, moveID uuid.U
...
+	if shipment.OriginalDeliveryDate == nil {
+		return MoveDatesSummary{}, errors.New("Shipment must have a OriginalDeliveryDate")
+	}
+	var mostCurrentDeliveryDate time.Time
+	if shipment.ActualDeliveryDate != nil {
+		mostCurrentDeliveryDate = *shipment.ActualDeliveryDate
+	} else if shipment.PmSurveyPlannedDeliveryDate != nil {
+		mostCurrentDeliveryDate = *shipment.PmSurveyPlannedDeliveryDate
+	} else {
+		mostCurrentDeliveryDate = *shipment.OriginalDeliveryDate
```
Comment (chrisgilmerproj) [inline pkg/handlers/internalapi/move_dates.go]: I think you might consider

```suggestion
} else if shipment.OriginalDeliveryDate != nil {
		mostCurrentDeliveryDate = *shipment.OriginalDeliveryDate
} else if shipment.EstimatedTransitDays != nil && shipment.RequestedPickupDate != nil {
		mostCurrentDeliveryDate = *shipment.RequestedPickupDate + *shipment.EstimatedTransitDays
}
```

File: pkg/handlers/internalapi/move_dates.go (hunk 1 above)
Comment (kimallen) [reply pkg/handlers/internalapi/move_dates.go]: yes

File: pkg/handlers/internalapi/move_dates.go (hunk 2 above)
Comment (kimallen) [reply pkg/handlers/internalapi/move_dates.go]: It just shows up blank, but I just changed it so that it says "TBD".  That seems more intentional and informative.

File: pkg/handlers/internalapi/move_dates.go (hunk 4 above)
Comment (kimallen) [reply pkg/handlers/internalapi/move_dates.go]: Those are needed when we only have the selected pickup date on the datepicker page in the wizard.  I added comments

File: pkg/handlers/internalapi/move_dates.go (hunk 5 above)
Comment (kimallen) [reply pkg/handlers/internalapi/move_dates.go]: I changed it to use the estimated delivery days (also the estimated pack days) to calcluate.  Wrote it a bit differently because of the need to account for holidays/weekends.
